src/main/java/com/github/manolo8/darkbot/core/api/ApiAdapter.java

```diff
@@ -34,32 +28,36 @@ public String readMemoryString(long address) {
 
     @Override
     public String readMemoryStringFallback(long address, String fallback) {
-        String str = stringCache.getUnchecked(address);
-        if (str == null) {
-            stringCache.invalidate(address);
-            return fallback;
-        }
-
-        if (str.isEmpty()) stringCache.invalidate(address);
-        return str;
+        String str = readMemoryStringInternal(address);
+        return str == null ? fallback : str;
     }
 
     public String readMemoryStringInternal(long address) {
+        int size = readMemoryInt(address + 32);
+        if (size == 0) return "";
+
+        //get string from cache and return if exists
+        String temp = stringCache.get(address);
+        if (temp != null) return temp;
```

Comment:
recheck if empty?

Reply:
empty cache values simply don't get into the cache, so that's fine